sympy/core/expr.py

```diff
@@ -1114,39 +1115,61 @@ def as_ordered_terms(self, order=None, data=False):
-            add_args = sorted(Add.make_args(self), key=key)
-            if (len(add_args) == 2
-                and isinstance(add_args[0], (Number, NumberSymbol))
-                and isinstance(add_args[1], Mul)):
-                mul_args = sorted(Mul.make_args(add_args[1]), key=key)
-                if (len(mul_args) == 2
-                    and isinstance(mul_args[0], Number)
-                    and add_args[0].is_positive
-                    and mul_args[0].is_negative):
-                    return add_args
-
         key, reverse = self._parse_order(order)
         terms, gens = self.as_terms()
 
-        if not any(term.is_Order for term, _ in terms):
-            ordered = sorted(terms, key=key, reverse=reverse)
-        else:
-            _terms, _order = [], []
-
-            for term, repr in terms:
-                if not term.is_Order:
-                    _terms.append((term, repr))
-                else:
-                    _order.append((term, repr))
+        _terms, _order = [], []
 
-            ordered = sorted(_terms, key=key, reverse=True) \
-                + sorted(_order, key=key, reverse=True)
+        for term, repr in terms:
+            if not term.is_Order:
+                _terms.append((term, repr))
+            else:
+                _order.append((term, repr))
+
+        if _order:
+            reverse = True
+        ordered = sorted(_terms, key=key, reverse=reverse) \
+            + sorted(_order, key=key, reverse=reverse)
+
+        # TOUCH UP
+        # (1) prefer y - x to -x + y but -1 + sqrt(2) to sqrt(2) - 1
+        def _has_radical(e):
+            from sympy.simplify.radsimp import denom
+            return any(denom(i.exp) != 1 for i in e.atoms(Pow))
+        def _powers(e):
+            p = Counter()
+            for f in Mul.make_args(e):
+                if f.is_number:
+                    continue
+                base, exp = f.as_base_exp()
+                p[base] += exp
+            return p
+        def _do_swap(a, b):
+            if _coeff_isneg(a) and not _coeff_isneg(b):
+                # don't do unless all bases in common
+                # have the same power
+                pa = _powers(a)
+                pb = _powers(b)
+                if all(pa[i] == pb[i] for i in set(pa) & set(pb)):
+                    return True
+            return False
+        if order is None and len(ordered) == 2 and self.is_Add:
+            a, b = [i[0] for i in ordered]
+            if  _do_swap(a, b):
+                ordered = ordered[::-1]
+                a, b = b, a
+            if _has_radical(a) and not _has_radical(b):
+                ordered = ordered[::-1]
+
+        # (2) always put I-containing terms last
+        n = len(ordered)
+        i = 0
+        while n:
+            if ordered[i][0].has(S.ImaginaryUnit):
+                ordered.append(ordered.pop(i))
+            else:
+                i += 1
+            n -= 1
```

Comment:
These rules are very complicated. I would aim to simplify things rather than make them more complicated.

Why is any of this needed?

I think that most users would just like an expression to print with its arguments in the same order as the args in the Add. Those are already sorted so why do we need to do anything else?

Reply:
The rules are few:
1. put terms with `I` at the end of an expression
2. for two-term Add, print `y - x` instead of `-x + y` unless doing so
    a. would put a radical term before a non-radical term
    b. would put a generator of a lower degree before a higher degree
    c. would violate rule 1
 
See the discussion in #16090 -- people don't want to see things as they appear in args. Anyone who does can print args; everyon else will print the object. Core cares about order because that is needed to make routines deterministic. But it doesn't demand as much as a person: it just needs them to be in unambiguous order without ties.